### ISSUE TITLE
Accessing instance data on shared_data blocks

### ISSUE THREAD

**Issue by rbngzlv** [inertiajs/inertia-rails#4]
Hi,

First of all thanks for the rails adapter for Inertia!

I'm planing to use the `shared_data` functionality to send the current user, locale, menu items ... but I think that is not possible right now, let me explain:

If I use the examples on the docs (https://inertiajs.com/shared-data#sharing-data) they work for synchronously and partially for the lazily case, because we don't have access to the controller instance inside the block, so we can't call the url helpers, flash, current_user, methods, etc, true?

In case that this is true:

Are you using another approach to send this data to Inertia? Or are you rendering the entire layout in Rails and only the content part with Inertia (not tested) ? Is this intended or can this be an improvement?

**Comment by BrandonShar** [comment]
For us personally, we're rendering things like the page header within the rails layout, but I would imagine your use case should still work.

You're saying that something like `current_user` doesn't work inside a block though? Does rails allow us to access the controller instance inside of a lambda instead of a block?

**Comment by rbngzlv** [comment]
Your approach rendering the page header o similar things in the rails layout is good because they doesn't get re-rendered in subsequent requests, right?

About the "issue" here, this works:

```ruby
class ActivityCategoriesController < ApplicationController
  def index
    puts current_user # I can access
  end
end
```

Any of this works:

```ruby
class ActivityCategoriesController < ApplicationController
  inertia_share do
    { user: current_user }
  end

 inertia_share user: lambda { current_user }
end
```

They raise:

```
undefined local variable or method `current_user' for ActivityCategoriesController:Class
```

I understand that this is because we are registering the block at the level class, I will try to investigate more about this.

And thank you for helping with this.

**Comment by BrandonShar** [comment]
Yes, that's exactly the pro to rendering them directly in the layout. Anything shared via `inertia_share` is going to be evaluated and sent in every inertia response. If this isn't data that's going to change from request to request, it's better to not have the extra overhead.

And thanks for the examples! That definitely makes sense (though I didn't expect it haha). I'm guessing we'll have to handle registering this function on the controller in a different way. Tagging @bknoles for some more rails knowledge.

**Comment by rbngzlv** [comment]
Jaja, don't worry. I also tested one thing that worked:

This block is executed inside the controller instance, so we have access to the controller instance with `self` and I can correctly call methods like `current_user` :

https://github.com/inertiajs/inertia-rails/blob/e51b6eaf961fc6f6946b93aaaca6674be7452988/lib/inertia.rb#L4-L13

So I tested passing the controller instance to the Inertia renderer class and execute the blocks with:

```ruby
def self.evaluated_blocks(controller, blocks)
  blocks.flat_map { |b| controller.instance_exec(&b) }.reduce(&:merge) || {}
 #                      ----------------------------
end
```

And it works, but I don't now if this is the best approach, or for example is preferable to execute the blocks inside the block registered as renderer (the first show in this comment) and pass the result merged with `props` to the render.

Another approach can be to have a function called `inertia_data` inside the controller and call it from the renderer, but we will lose the possibility to call `inertia_share` multiple times without fear and take care to call super if we override the function in a child controller.

So waiting for @bknoles comments and trying your approach of rendering the layout directly in Rails. :smile: 

Thanks you all.

**Comment by bknoles** [comment]
@rbngzlv 

thanks for the code examples and for bringing this up! let me make sure I understand this correctly. I'll summarize in bullet points (because i ❤️bullet points)

* In the current version of Inertia Rails, the `Inertia.share` controller method doesn't have access to the controller instance, and therefore helper methods (like `current_user`, etc) are not available
* You have a created working solution for `Inertia.share` with a block by explicitly passing the controller reference to the custom renderer and passing it along (the same way it currently passes request, response, render_method, if I read this correctly)
* The working solution doesn't work if a lambda is used instead of a block

does that correctly describe the state of things?

**Comment by rbngzlv** [comment]
Hi @bknoles ,

> does that correctly describe the state of things?

Yes. You can see a diff here: https://github.com/inertiajs/inertia-rails/compare/master...rbngzlv:controller_instance_exec

**Comment by bknoles** [comment]
thanks!

i will take a look at this either today or tomorrow. weird that lambdas and blocks would behave differently in this situation.

in any case, the basic use case (using controller methods in the shared data blocks) seems like an obvious one that the gem should support.

@BrandonShar how do you feel about merging in Ruben's changes before figuring out why lambdas behave differently?

**Comment by BrandonShar** [comment]
They look good to me overall, but it would be easier to comment in a PR.

@rbngzlv do you feel the code is solid enough to PR it?

**Comment by rbngzlv** [comment]
Hi, I'm using it to develop an application right now and it is working well for now. 

I'm going to open the PR so we can discuss further on it.

**Comment by bknoles** [comment]
I checked this out in my test app I've been using to verify Inertia stuff, and I see what's going on.

When we share data, we treat it as either a plain hash OR as a callable block:

```ruby
module InertiaRails
  module Controller
    extend ActiveSupport::Concern

    module ClassMethods
      def inertia_share(**args, &block)
        InertiaRails.share(args) if args
        InertiaRails.share_block(block) if block
      end
    end
  end
end
```

The way we have the documentation specified, this:

```
inertia_share user_count: lambda { User.count }
```

will be treated as a plain hash... so, that lambda will not execute in the context of the controller that @rbngzlv is passing into the `evaluated_blocks` method.

In order to support that syntax properly, I think we need an `evaluated_plain_data` method that also receives the controller reference (similar to `evaluated_blocks`) that will check whether something is a lambda and needs to be evaluated.

Pretty simple for the documentation example, but we'll probably also want to traverse the hash in case it's got nested data like this:

```
inertia_share general_data: {
  app_name: 'The best app',
  user: lambda { current_user },
}
```

**Comment by rbngzlv** [comment]
@bknoles @BrandonShar updated the pull request to also execute lambdas inside the controller instance. Your two examples (nested and not nested) are working fine.

**Comment by bknoles** [comment]
Merged! We released `1.0.1` with your changes @rbngzlv 

Thanks for taking this on!
